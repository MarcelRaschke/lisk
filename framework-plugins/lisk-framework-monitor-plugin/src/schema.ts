/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
export const postBlockEventSchema = {
	$id: 'monitor/postBlockEvent',
	type: 'object',
	required: ['block'],
	properties: {
		block: {
			type: 'string',
			format: 'hex',
		},
	},
};

export const transactionAnnouncementSchema = {
	$id: 'monitor/transactionAnnouncement',
	type: 'object',
	required: ['transactionIds'],
	properties: {
		transactionIds: {
			type: 'array',
			items: {
				type: 'string',
				format: 'hex',
			},
		},
	},
};
