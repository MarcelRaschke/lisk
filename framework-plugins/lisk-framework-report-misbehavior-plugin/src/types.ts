/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export interface Options {
	version: string;
	readonly dataPath: string;
	readonly encryptedPassphrase: string;
	readonly defaultPassword: string;
	readonly port: number;
	readonly whiteList: ReadonlyArray<string>;
	readonly cors: {
		readonly origin: string;
		readonly methods: string[];
	};
	readonly limits: {
		readonly max: number;
		readonly delayMs: number;
		readonly delayAfter: number;
		readonly windowMs: number;
		readonly headersTimeout: number;
		readonly serverSetTimeout: number;
	};
	readonly fee: number;
	readonly clearBlockHeadersInterval: number;
}

export interface State {
	publicKey?: Buffer;
	passphrase?: string;
	currentHeight: number;
}
