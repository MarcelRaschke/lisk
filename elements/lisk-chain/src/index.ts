/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED } from './constants';

const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK, EVENT_VALIDATORS_CHANGED };

export { events };
export { Chain } from './chain';
export { Transaction, TransactionInput, transactionSchema, calculateMinFee } from './transaction';
export {
	blockHeaderAssetSchema,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	validatorsSchema,
	getGenesisBlockHeaderAssetSchema,
	stateDiffSchema,
	getRegisteredBlockAssetSchema,
} from './schema';
export {
	CONSENSUS_STATE_VALIDATORS_KEY,
	CONSENSUS_STATE_FINALIZED_HEIGHT_KEY,
	TAG_BLOCK_HEADER,
	TAG_TRANSACTION,
} from './constants';
export type {
	Account,
	AccountDefaultProps,
	RawBlock,
	RawBlockHeader,
	GenesisBlock,
	GenesisBlockHeader,
	Block,
	BlockHeader,
	BlockHeaderAsset,
	Validator,
	AccountSchema,
} from './types';
export { Slots } from './slots';
export { readGenesisBlockJSON, getValidators, getAccountSchemaWithDefault } from './utils';
export * as testing from './testing';

export type { StateStore } from './state_store';
export type { DataAccess } from './data_access';
