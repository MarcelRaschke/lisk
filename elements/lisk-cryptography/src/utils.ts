/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

export const readBit = (buf: Buffer, bit: number): boolean => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	// eslint-disable-next-line no-bitwise
	return (buf[byteIndex] >> bitIndex) % 2 === 1;
};

export const writeBit = (buf: Buffer, bit: number, val: boolean): void => {
	const byteIndex = Math.floor(bit / 8);
	const bitIndex = bit % 8;

	if (val) {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] |= 1 << bitIndex;
	} else {
		// eslint-disable-next-line no-bitwise, no-param-reassign
		buf[byteIndex] &= ~(1 << bitIndex);
	}
};
