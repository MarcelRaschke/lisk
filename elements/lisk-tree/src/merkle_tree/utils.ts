/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
/* eslint-disable no-bitwise */

import { hash } from '@liskhq/lisk-cryptography';
import { BRANCH_PREFIX, LEAF_PREFIX } from './constants';
import { NodeLocation, NodeSide, MerkleRootInfo } from './types';

export const isLeaf = (value: Buffer): boolean => value[0] === LEAF_PREFIX[0];

export const generateHash = (prefix: Buffer, leftHash: Buffer, rightHash: Buffer): Buffer =>
	hash(
		Buffer.concat(
			[prefix, leftHash, rightHash],
			prefix.length + leftHash.length + rightHash.length,
		),
	);

export const getMaxIdxAtLayer = (layer: number, size: number): number => {
	let [max, r] = [size, 0];
	for (let i = 0; i < layer; i += 1) {
		[max, r] = [[Math.floor, Math.ceil][r % 2](max / 2), r + (max % 2)];
	}
	return max;
};

export const getLayerStructure = (size: number): number[] => {
	const structure = [];
	for (let i = 0; i <= Math.ceil(Math.log2(size)); i += 1) {
		structure.push(getMaxIdxAtLayer(i, size));
	}

	return structure;
};

export const getBinaryString = (num: number, length: number): Buffer => {
	if (length === 0) {
		return Buffer.alloc(0);
	}
	let binaryString = num.toString(2);
	while (binaryString.length < length) {
		binaryString = `0${binaryString}`;
	}

	return Buffer.from(binaryString, 'utf8');
};

export const getBinary = (num: number, length: number): number[] => {
	const binaryString = getBinaryString(num, length).toString('utf8');

	return binaryString.split('').map(d => parseInt(d, 10));
};

export const getRightSiblingInfo = (
	nodeIndex: number,
	layerIndex: number,
	size: number,
): NodeLocation | undefined => {
	const structure = getLayerStructure(size);
	let siblingNodeIndex = ((nodeIndex >>> 1) << 1) + ((nodeIndex + 1) % 2);
	let siblingLayerIndex = layerIndex;
	while (siblingNodeIndex >= structure[siblingLayerIndex] && siblingLayerIndex > 0) {
		siblingNodeIndex <<= 1;
		siblingLayerIndex -= 1;
	}
	if (siblingLayerIndex === 0 && siblingNodeIndex >= size) {
		return undefined;
	}
	return {
		nodeIndex: siblingNodeIndex,
		layerIndex: siblingLayerIndex,
	};
};

export const getPairLocation = (nodeInfo: {
	layerIndex: number;
	nodeIndex: number;
	size: number;
}): NodeLocation => {
	const { layerIndex, nodeIndex, size } = nodeInfo;
	const treeHeight = Math.ceil(Math.log2(size)) + 1;
	const layerStructure = getLayerStructure(size);
	const numberOfNodesInLayer = layerStructure[layerIndex];
	const binary = getBinary(nodeIndex, treeHeight - layerIndex);
	const side = [NodeSide.LEFT, NodeSide.RIGHT][binary[binary.length - 1]];
	const pairSide = side === NodeSide.LEFT ? NodeSide.RIGHT : NodeSide.LEFT;

	// If queried node is root, provide root node location
	if (layerIndex + 1 === treeHeight) {
		return { layerIndex: treeHeight - 1, nodeIndex: 0 };
	}
	// If node is left node not last element in the layer
	if (side === NodeSide.LEFT && nodeIndex < numberOfNodesInLayer - 1) {
		const pairNodeIndex = nodeIndex + 1;
		return { layerIndex, nodeIndex: pairNodeIndex, side: pairSide };
	}
	// If node is right node AND (not last element in layer OR last element in the layer with even # of nodes)
	if (
		side === NodeSide.RIGHT &&
		((numberOfNodesInLayer % 2 === 0 && nodeIndex === numberOfNodesInLayer - 1) ||
			(nodeIndex < numberOfNodesInLayer - 1 && nodeIndex < numberOfNodesInLayer - 1))
	) {
		const pairNodeIndex = nodeIndex - 1;
		return { layerIndex, nodeIndex: pairNodeIndex, side: pairSide };
	}
	// Otherwise find next odd numbered layer
	let currentLayer = layerIndex;
	// Get direction to traverse tree
	const numOfOddLayers = layerStructure
		.slice(0, currentLayer)
		.filter(num => num % 2 !== 0)
		.reduce((acc, val) => acc + val, 0);
	const direction = numOfOddLayers % 2 === 0 ? 1 : -1;
	let pairLocation;
	currentLayer += direction;
	while (currentLayer >= 0 && currentLayer <= treeHeight - 1) {
		if (layerStructure[currentLayer] % 2 !== 0) {
			const pairNodeIndex =
				direction === 1
					? layerStructure[currentLayer] + direction * -1
					: layerStructure[currentLayer] - direction * -1;
			pairLocation = {
				layerIndex: currentLayer,
				nodeIndex: pairNodeIndex,
				side: direction === -1 ? NodeSide.RIGHT : NodeSide.LEFT,
			};
			break;
		}
		currentLayer += direction;
	}

	return pairLocation as NodeLocation;
};

export const calculateMerkleRoot = ({ value, appendPath, size }: MerkleRootInfo) => {
	// 1. Calculate the new root

	// Add prefix to value
	const leafValueWithPrefix = Buffer.concat(
		[LEAF_PREFIX, value],
		LEAF_PREFIX.length + value.length,
	);

	// Set current hash to hash of value
	const newLeafHash = hash(leafValueWithPrefix);
	let currentHash = newLeafHash;

	// Count the 1's in binaryLength
	let count = 0;

	const binaryLength = size.toString(2);

	for (let i = 0; i < binaryLength.length; i += 1) {
		// Loop the binaryLength from the right
		// The right-most digits correspond to lower layers in the tree
		if ((size >> i) & 1) {
			const siblingHash = appendPath[count];
			currentHash = generateHash(BRANCH_PREFIX, siblingHash, currentHash);
			count += 1;
		}
	}

	const newRoot = currentHash;

	// 2. Update the append path
	let subTreeIndex;
	const treeHeight = Math.ceil(Math.log2(size)) + 1;

	for (subTreeIndex = 0; subTreeIndex < treeHeight; subTreeIndex += 1) {
		if (!((size >> subTreeIndex) & 1)) {
			break;
		}
	}

	const currentAppendPath = appendPath.slice(0);
	currentHash = newLeafHash;
	const splicedPath = currentAppendPath.splice(subTreeIndex);
	for (const sibling of currentAppendPath) {
		currentHash = generateHash(BRANCH_PREFIX, sibling, currentHash);
	}
	const newAppendPath = [currentHash].concat(splicedPath);

	return { root: newRoot, appendPath: newAppendPath, size: size + 1 };
};
