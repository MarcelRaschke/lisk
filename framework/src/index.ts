/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export {
	Transaction,
	transactionSchema,
	readGenesisBlockJSON,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	getAccountSchemaWithDefault,
	getGenesisBlockHeaderAssetSchema,
	blockHeaderAssetSchema,
} from '@liskhq/lisk-chain';
export {
	BaseModule,
	BaseAsset,
	TokenModule,
	TokenTransferAsset,
	SequenceModule,
	KeysModule,
	KeysRegisterAsset,
	DPoSModule,
	DPoSRegisterAsset,
	DPoSVoteAsset,
	DPoSUnlockAsset,
	DPoSPoMAsset,
} from './modules';
export { Application } from './application';
export { systemDirs } from './system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsDefinition, EventCallback } from './controller/event';
export type { ActionsDefinition, ActionHandler } from './controller/action';
export * as testing from './testing';
export * from './types';
