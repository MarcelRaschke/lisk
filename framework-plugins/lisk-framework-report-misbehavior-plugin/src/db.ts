/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { formatInt, KVStore, getFirstPrefix, getLastPrefix } from '@liskhq/lisk-db';
import { codec } from '@liskhq/lisk-codec';
import { RawBlockHeader, BlockHeader } from '@liskhq/lisk-chain';
import { areHeadersContradicting } from '@liskhq/lisk-bft';
import * as os from 'os';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { RegisteredSchema } from 'lisk-framework';
import { hash } from '@liskhq/lisk-cryptography';

export const blockHeadersSchema = {
	$id: 'lisk/reportMisbehavior/blockHeaders',
	type: 'object',
	required: ['blockHeaders'],
	properties: {
		blockHeaders: {
			type: 'array',
			fieldNumber: 1,
			items: {
				dataType: 'bytes',
			},
		},
	},
};

export interface BlockHeaderAsset {
	readonly seedReveal: Buffer;
	readonly maxHeightPreviouslyForged: number;
	readonly maxHeightPrevoted: number;
}

interface BlockHeaders {
	readonly blockHeaders: Buffer[];
}

export const getDBInstance = async (
	dataPath: string,
	dbName = 'lisk-framework-report-misbehavior-plugin.db',
): Promise<KVStore> => {
	const dirPath = join(dataPath.replace('~', os.homedir()), 'plugins/data', dbName);
	await ensureDir(dirPath);

	return new KVStore(dirPath);
};

export const getBlockHeaders = async (
	db: KVStore,
	dbKeyBlockHeader: string,
): Promise<BlockHeaders> => {
	try {
		const encodedBlockHeaders = await db.get(dbKeyBlockHeader);
		return codec.decode<BlockHeaders>(blockHeadersSchema, encodedBlockHeaders);
	} catch (error) {
		return { blockHeaders: [] as Buffer[] };
	}
};

export const decodeBlockHeader = (encodedHeader: Buffer, schema: RegisteredSchema): BlockHeader => {
	const id = hash(encodedHeader);
	const blockHeader = codec.decode<RawBlockHeader>(schema.blockHeader, encodedHeader);
	const assetSchema = schema.blockHeadersAssets[blockHeader.version];
	const asset = codec.decode<BlockHeaderAsset>(assetSchema, blockHeader.asset);
	return {
		...blockHeader,
		asset,
		id,
	};
};

export const saveBlockHeaders = async (
	db: KVStore,
	schemas: RegisteredSchema,
	header: Buffer,
): Promise<boolean> => {
	const blockId = hash(header);
	const { generatorPublicKey, height } = codec.decode<RawBlockHeader>(schemas.blockHeader, header);
	const dbKey = `${generatorPublicKey.toString('binary')}:${formatInt(height)}`;
	const { blockHeaders } = await getBlockHeaders(db, dbKey);

	if (!blockHeaders.find(blockHeader => hash(blockHeader).equals(blockId))) {
		await db.put(
			dbKey,
			codec.encode(blockHeadersSchema, {
				blockHeaders: [...blockHeaders, header],
			}),
		);
		return true;
	}
	return false;
};

type IteratableStream = NodeJS.ReadableStream & { destroy: (err?: Error) => void };

export const getContradictingBlockHeader = async (
	db: KVStore,
	blockHeader: BlockHeader,
	schemas: RegisteredSchema,
): Promise<BlockHeader | undefined> =>
	new Promise((resolve, reject) => {
		const stream = db.createReadStream({
			gte: getFirstPrefix(blockHeader.generatorPublicKey.toString('binary')),
			lte: getLastPrefix(blockHeader.generatorPublicKey.toString('binary')),
		}) as IteratableStream;
		stream
			.on('data', ({ value }: { value: Buffer }) => {
				const { blockHeaders } = codec.decode<BlockHeaders>(blockHeadersSchema, value);
				for (const encodedHeader of blockHeaders) {
					const decodedBlockHeader = decodeBlockHeader(encodedHeader, schemas);
					if (areHeadersContradicting(blockHeader, decodedBlockHeader)) {
						stream.destroy();
						resolve(decodedBlockHeader);
					}
				}
			})
			.on('error', error => {
				reject(error);
			})
			.on('end', () => {
				resolve(undefined);
			});
	});

export const clearBlockHeaders = async (
	db: KVStore,
	schemas: RegisteredSchema,
	currentHeight: number,
): Promise<void> => {
	const keys = await new Promise<string[]>((resolve, reject) => {
		const stream = db.createReadStream() as IteratableStream;
		const res: string[] = [];
		stream
			.on('data', ({ key, value }: { key: string; value: Buffer }) => {
				const { blockHeaders } = codec.decode<BlockHeaders>(blockHeadersSchema, value);
				for (const encodedHeader of blockHeaders) {
					const decodedBlockHeader = decodeBlockHeader(encodedHeader, schemas);
					if (decodedBlockHeader.height < currentHeight - 260000) {
						res.push(key);
					}
				}
			})
			.on('error', error => {
				reject(error);
			})
			.on('end', () => {
				resolve(res);
			});
	});
	const batch = db.batch();
	for (const k of keys) {
		batch.del(k);
	}
	await batch.write();
};
