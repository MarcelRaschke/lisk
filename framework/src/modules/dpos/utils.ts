/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { verifyData } from '@liskhq/lisk-cryptography';
import { Account } from '@liskhq/lisk-chain';
import { DPOSAccountProps, UnlockingAccountAsset } from './types';
import {
	PUNISHMENT_PERIOD,
	SELF_VOTE_PUNISH_TIME,
	VOTER_PUNISH_TIME,
	WAIT_TIME_SELF_VOTE,
	WAIT_TIME_VOTE,
} from './constants';

export const sortUnlocking = (unlocks: UnlockingAccountAsset[]): void => {
	unlocks.sort((a, b) => {
		if (!a.delegateAddress.equals(b.delegateAddress)) {
			return a.delegateAddress.compare(b.delegateAddress);
		}
		if (a.unvoteHeight !== b.unvoteHeight) {
			return b.unvoteHeight - a.unvoteHeight;
		}
		const diff = b.amount - a.amount;
		if (diff > BigInt(0)) {
			return 1;
		}
		if (diff < BigInt(0)) {
			return -1;
		}

		return 0;
	});
};

export const getMinPunishedHeight = (
	sender: Account<DPOSAccountProps>,
	delegate: Account<DPOSAccountProps>,
): number => {
	if (delegate.dpos.delegate.pomHeights.length === 0) {
		return 0;
	}

	const lastPomHeight = Math.max(...delegate.dpos.delegate.pomHeights);

	// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0024.md#update-to-validity-of-unlock-transaction
	return sender.address.equals(delegate.address)
		? lastPomHeight + SELF_VOTE_PUNISH_TIME
		: lastPomHeight + VOTER_PUNISH_TIME;
};

export const getPunishmentPeriod = (
	sender: Account<DPOSAccountProps>,
	delegateAccount: Account<DPOSAccountProps>,
	lastBlockHeight: number,
): number => {
	const currentHeight = lastBlockHeight + 1;
	const minPunishedHeight = getMinPunishedHeight(sender, delegateAccount);
	const remainingBlocks = minPunishedHeight - currentHeight;

	return remainingBlocks < 0 ? 0 : remainingBlocks;
};

export const getMinWaitingHeight = (
	senderAddress: Buffer,
	delegateAddress: Buffer,
	unlockObject: UnlockingAccountAsset,
): number =>
	unlockObject.unvoteHeight +
	(senderAddress.equals(delegateAddress) ? WAIT_TIME_SELF_VOTE : WAIT_TIME_VOTE);

export const getWaitingPeriod = (
	senderAddress: Buffer,
	delegateAddress: Buffer,
	lastBlockHeight: number,
	unlockObject: UnlockingAccountAsset,
): number => {
	const currentHeight = lastBlockHeight + 1;
	const minWaitingHeight = getMinWaitingHeight(senderAddress, delegateAddress, unlockObject);
	const remainingBlocks = minWaitingHeight - currentHeight;

	return remainingBlocks < 0 ? 0 : remainingBlocks;
};

export const isNullCharacterIncluded = (input: string): boolean =>
	new RegExp(/\\0|\\u0000|\\x00/).test(input);

export const isUsername = (username: string): boolean => {
	if (isNullCharacterIncluded(username)) {
		return false;
	}

	if (username !== username.trim().toLowerCase()) {
		return false;
	}

	return /^[a-z0-9!@$&_.]+$/g.test(username);
};

export const validateSignature = (
	tag: string,
	networkIdentifier: Buffer,
	publicKey: Buffer,
	signature: Buffer,
	bytes: Buffer,
): boolean => verifyData(tag, networkIdentifier, bytes, signature, publicKey);

export const isCurrentlyPunished = (height: number, pomHeights: ReadonlyArray<number>): boolean => {
	if (pomHeights.length === 0) {
		return false;
	}
	const lastPomHeight = Math.max(...pomHeights);
	if (height - lastPomHeight < PUNISHMENT_PERIOD) {
		return true;
	}

	return false;
};
