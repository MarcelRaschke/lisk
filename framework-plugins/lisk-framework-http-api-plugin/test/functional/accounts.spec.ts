/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { testing, PartialApplicationConfig } from 'lisk-framework';
import axios from 'axios';

import { callNetwork, getURL } from './utils/application';
import { HTTPAPIPlugin } from '../../src/http_api_plugin';

describe('Account endpoint', () => {
	let appEnv: testing.ApplicationEnv;
	const label = 'account_http_functional';

	beforeAll(async () => {
		const rootPath = '~/.lisk/http-plugin';
		const config = {
			rootPath,
			label,
		} as PartialApplicationConfig;

		appEnv = testing.createDefaultApplicationEnv({
			config,
			plugins: [HTTPAPIPlugin],
		});
		await appEnv.startApplication();
	});

	afterAll(async () => {
		// eslint-disable-next-line @typescript-eslint/no-empty-function
		jest.spyOn(process, 'exit').mockImplementation((() => {}) as never);
		await appEnv.stopApplication();
	});

	describe('/api/accounts', () => {
		it('should respond with account when account found in db', async () => {
			const result = await axios.get(
				getURL('/api/accounts/9d0149b0962d44bfc08a9f64d5afceb6281d7fb5'),
			);
			expect(result.data).toMatchSnapshot();
			expect(result.status).toBe(200);
		});

		it('should respond with 404 and error message when account not found in db', async () => {
			const { response, status } = await callNetwork(
				axios.get(getURL('/api/accounts/9d0149b0962d44bfc08a9d24d5afceb6281d7fb5')),
			);

			expect(status).toBe(404);
			expect(response).toEqual({
				errors: [
					{
						message:
							"Account with address '9d0149b0962d44bfc08a9d24d5afceb6281d7fb5' was not found",
					},
				],
			});
		});

		it('should respond with 400 and error message when address param is not hex', async () => {
			const { response, status } = await callNetwork(axios.get(getURL('/api/accounts/-nein-no')));

			expect(status).toBe(400);
			expect(response).toEqual({
				errors: [{ message: 'The Address parameter should be a hex string.' }],
			});
		});
	});
});
