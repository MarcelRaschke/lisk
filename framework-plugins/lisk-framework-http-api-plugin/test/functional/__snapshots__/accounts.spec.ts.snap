// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account endpoint /api/accounts should respond with account when account found in db 1`] = `
Object {
  "data": Object {
    "address": "9d0149b0962d44bfc08a9f64d5afceb6281d7fb5",
    "dpos": Object {
      "delegate": Object {
        "consecutiveMissedBlocks": 0,
        "isBanned": false,
        "lastForgedHeight": 0,
        "pomHeights": Array [],
        "totalVotesReceived": "0",
        "username": "delegate_4",
      },
      "sentVotes": Array [],
      "unlocking": Array [],
    },
    "keys": Object {
      "mandatoryKeys": Array [],
      "numberOfSignatures": 0,
      "optionalKeys": Array [],
    },
    "sequence": Object {
      "nonce": "0",
    },
    "token": Object {
      "balance": "0",
    },
  },
  "meta": Object {},
}
`;
